src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+     */
+    private abstract static class Ref {
+
+        private final int depth;
+        private final boolean direct;
+
+        Ref(int depth, boolean direct) {
+            this.depth = depth;
+            this.direct = direct;
+        }
+
+        public int getDepth() {
+            return this.depth;
+        }
+
+        public boolean isDirect() {
+            return this.direct;
+        }
+
+        @Override
+        public int hashCode() {
+            return this.getClass().hashCode()
+                ^ Integer.hashCode(this.depth)
+                ^ Boolean.hashCode(this.direct);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj == null || obj.getClass() != this.getClass())
+                return false;
+            final Ref that = (Ref)obj;
+            return this.depth == that.depth
+              && this.direct == that.direct;
+        }
+
+        @Override
+        public String toString() {
+            final ArrayList<String> properties = new ArrayList<>();
+            this.addProperties(properties);
+            return this.getClass().getSimpleName()
+              + "[" + properties.stream().collect(Collectors.joining(",")) + "]";
+        }
+
+        protected void addProperties(ArrayList<String> properties) {
+            properties.add("depth=" + this.depth);
+            properties.add(this.direct ? "direct" : "indirect");
+        }
+    }
+
+    /** A reference from the current 'this' instance.
+     */
+    private static class ThisRef extends Ref {
+
+        ThisRef(boolean direct) {
+            super(0, direct);
+        }
+
+        public static ThisRef direct() {
+            return new ThisRef(true);
+        }
+
+        public static ThisRef indirect() {
+            return new ThisRef(false);
+        }
+    }
+
+    /** A reference from the current outer 'this' instance.
+     */
+    private static class OuterRef extends Ref {
+
+        OuterRef(boolean direct) {
+            super(0, direct);
+        }
+
+        public static OuterRef direct() {
+            return new OuterRef(true);
+        }
+
+        public static OuterRef indirect() {
+            return new OuterRef(false);
+        }
+    }
+
+    /** A reference from the expression that was just evaluated.
+     *  In other words, a reference that's sitting on top of the stack.
+     */
+    private static class ExprRef extends Ref {
+
+        ExprRef(int depth, boolean direct) {
+            super(depth, direct);
+        }
+
+        public static ExprRef direct(int depth) {
+            return new ExprRef(depth, true);
+        }
+
+        public static ExprRef indirect(int depth) {
+            return new ExprRef(depth, false);
+        }
+    }
+
+    /** A reference from the return value of the current method being "invoked".
+     */
+    private static class ReturnRef extends Ref {
```

Comment:
Isn't this just an ExprRef? This might also be related with the fact that we deal with return values in different ways than with e.g. values returned from a nested scope (where we just pop, and then copy all pending expression to the outer depth).

Reply:
> This might also be related with the fact that we deal with return values in different ways than with e.g. values returned from a nested scope (where we just pop, and then copy all pending expression to the outer depth).

Yes, a method return value that represents a reference is an `ExprRef` just before the `return` is actually executed, but then it turns into a `ReturnRef`.

A `ReturnRef` is really just a yellow sticky note that reminds us "Hey whenever you finish 'executing' this method, remember that there is a reference in its return value".

> This might also be related with the fact that we deal with return values in different ways than with e.g. values returned from a nested scope (where we just pop, and then copy all pending expression to the outer depth).

Exactly.